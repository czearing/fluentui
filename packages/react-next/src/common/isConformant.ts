import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(
  testInfo: Omit<IsConformantOptions, 'componentPath' | 'packageVersion'> & { componentPath?: string },
) {
  const defaultOptions = {
    disabledTests: ['has-docblock', 'kebab-aria-attributes'],
    componentPath: module!.parent!.filename.replace('.test', ''),
    packageVersion: '@fluentui/react-next',
  };

  baseIsConformant(defaultOptions, testInfo);
}
