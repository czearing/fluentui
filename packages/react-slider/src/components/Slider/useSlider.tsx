import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useControllableValue } from '@fluentui/react-utilities';
import {
  getCode,
  ArrowDownKey,
  ArrowUpKey,
  ArrowLeftKey,
  ArrowRightKey,
  PageDownKey,
  PageUpKey,
  HomeKey,
  EndKey,
} from '@fluentui/keyboard-key';
import { on } from '@fluentui/utilities';
import { SliderProps, SliderShorthandProps, SliderState, DragChangeEvent } from './Slider.types';
import { clamp } from './utils/clamp';
import { getPercent } from './utils/getPercent';

/**
 * Array of all shorthand properties listed in SliderShorthandProps
 */
export const sliderShorthandProps: SliderShorthandProps[] = ['thumb', 'rail', 'track'];

const mergeProps = makeMergeProps<SliderState>({ deepMerge: sliderShorthandProps });

/**
 * Create the state required to render Slider.
 *
 * The returned state can be modified with hooks such as useSliderStyles,
 * before being passed to renderSlider.
 *
 * @param props - props from this instance of Slider
 * @param ref - reference to root HTMLElement of Slider
 * @param defaultProps - (optional) default prop values provided by the implementing type
 */
export const useSlider = (props: SliderProps, ref: React.Ref<HTMLElement>, defaultProps?: SliderProps): SliderState => {
  const { value, defaultValue = 0, min = 0, max = 10, step = 1, ariaValueText, onChange } = props;

  const [currentValue, setCurrentValue] = useControllableValue(
    value,
    clamp(defaultValue, min, max),
    (ev: DragChangeEvent, val: number) => onChange?.(val, ev),
  );

  const railRef = React.useRef(null);
  const disposables = React.useRef<(() => void)[]>([]);

  /**
   * Updates the `currentValue` to the new `incomingValue` and clamps it.
   *
   * @param ev
   * @param incomingValue
   */
  const updateValue = React.useCallback(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (ev: any, incomingValue: number): void => {
      setCurrentValue(clamp(incomingValue, min, max), ev);
    },
    [max, min, setCurrentValue],
  );

  /**
   * Gets the current position of the `thumb` based on the cursor position.
   */
  const getPosition = (ev: DragChangeEvent): number => {
    let currentPosition = 0;

    switch (ev.type) {
      case 'mousedown':
      case 'mousemove':
        currentPosition = (ev as MouseEvent).clientX;
        break;
      case 'touchstart':
      case 'touchmove':
        currentPosition = (ev as TouchEvent).touches[0].clientX;
        break;
    }

    return currentPosition;
  };

  /**
   * Calculates the `step` position based off of a `Mouse` or `Touch` event.
   */
  const calculateSteps = (ev: DragChangeEvent): number => {
    const currentBounds = railRef.current.getBoundingClientRect();
    const size = currentBounds.left;
    const position = currentBounds.width;

    const totalSteps = (max - min) / step;
    const stepLength: number = size / totalSteps;
    const thumbPosition = getPosition(ev);
    const distance = thumbPosition - position;

    return distance / stepLength;
  };

  const onDrag = (ev: DragChangeEvent): void => {
    updateValue(ev, min + step * calculateSteps(ev));
  };

  const onThumbReleased = (): void => {
    disposables.current.forEach(dispose => dispose());
    disposables.current = [];
  };

  const onThumbPressed = (ev: React.MouseEvent | React.TouchEvent): void => {
    if (ev.type === 'mousedown') {
      disposables.current.push(
        on(window, 'mousemove', onDrag as (ev: Event) => void, true),
        on(window, 'mouseup', onThumbReleased, true),
      );
    } else if (ev.type === 'touchstart') {
      disposables.current.push(
        on(window, 'touchmove', onDrag as (ev: Event) => void, true),
        on(window, 'touchend', onThumbReleased, true),
      );
    }

    onDrag(ev);
  };

  const onKeyDown = React.useCallback(
    (ev: React.KeyboardEvent): void => {
      const key = getCode(ev);

      if (ev.shiftKey) {
        if (key === ArrowDownKey || key === ArrowLeftKey) {
          updateValue(ev, currentValue - step * 10);
          return;
        } else if (key === ArrowUpKey || key === ArrowRightKey) {
          updateValue(ev, currentValue + step * 10);
          return;
        }
      } else if (key === ArrowDownKey || key === ArrowLeftKey) {
        updateValue(ev, currentValue - step);
        return;
      } else if (key === ArrowUpKey || key === ArrowRightKey) {
        updateValue(ev, currentValue + step);
        return;
      } else {
        switch (getCode(ev)) {
          case PageDownKey:
            updateValue(ev, currentValue - step * 10);
            break;
          case PageUpKey:
            updateValue(ev, currentValue + step * 10);
            break;
          case HomeKey:
            updateValue(ev, min);
            break;
          case EndKey:
            updateValue(ev, max);
            break;
        }
      }
    },
    [currentValue, max, min, step, updateValue],
  );

  React.useEffect(() => {
    if (value) {
      const clampedValue = clamp(value, min, max);
      if (clampedValue !== value) {
        setCurrentValue(clampedValue);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps -- Should only run on mount to clamp value
  }, []);

  const valuePercent = getPercent(currentValue, min, max);

  const rootProps: Partial<SliderState> = {
    className: 'ms-Slider-root',
    onMouseDown: onThumbPressed,
    onTouchStart: onThumbPressed,
    onKeyDown: onKeyDown,
  };

  const railProps: React.HTMLAttributes<HTMLDivElement> & React.RefAttributes<HTMLDivElement> = {
    className: 'ms-Slider-rail',
    ref: railRef,
  };

  const trackProps: React.HTMLAttributes<HTMLDivElement> = {
    className: 'ms-Slider-track',
    style: { width: valuePercent },
  };

  const thumbProps: React.HTMLAttributes<HTMLDivElement> = {
    className: 'ms-Slider-thumb',
    style: { width: valuePercent },
    tabIndex: 0,
    role: 'slider',
    'aria-valuemin': min,
    'aria-valuemax': max,
    'aria-valuenow': Math.floor(currentValue),
    'aria-valuetext': ariaValueText ? ariaValueText(currentValue) : currentValue.toString(),
  };

  const state = mergeProps(
    {
      ref,
      as: 'div',
      track: { as: 'div', children: null, ...trackProps },
      rail: { as: 'div', children: null, ...railProps },
      thumb: { as: 'div', children: null, ...thumbProps },
      ...rootProps,
    },
    defaultProps && resolveShorthandProps(defaultProps, sliderShorthandProps),
    resolveShorthandProps(props, sliderShorthandProps),
  );

  return state;
};
