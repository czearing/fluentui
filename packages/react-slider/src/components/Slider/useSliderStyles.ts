import { makeStyles, mergeClasses } from '@fluentui/react-make-styles';
import { HighContrastSelector } from '@fluentui/style-utilities';
import { createFocusIndicatorStyleRule } from '@fluentui/react-tabster';
import { SliderState } from './Slider.types';

/**
 * Styles for the root slot
 */
const useRootStyles = makeStyles({
  root: theme => ({
    userSelect: 'none',
    position: 'relative',
    display: 'inline-flex',
    height: '25px',
    width: '280px',
    alignItems: 'center',
  }),

  hover: theme => ({
    ':hover .ms-Slider-thumb ': {
      background: '#0078D4',

      [HighContrastSelector]: {
        forcedColorAdjust: 'none',
      },
    },

    ':hover .ms-Slider-track': {
      background: '#0078D4',

      [HighContrastSelector]: {
        forcedColorAdjust: 'none',
      },
    },
  }),

  focusWithin: theme => ({
    ':focus-within .ms-Slider-track': {
      position: 'aboslute',
      background: '#0078D4',

      [HighContrastSelector]: {
        forcedColorAdjust: 'none',
      },
    },

    ':focus-within .ms-Slider-thumb': {
      background: '#0078D4',

      [HighContrastSelector]: {
        background: 'GrayText',
      },
    },
  }),

  activation: theme => ({
    ':active .ms-Slider-thumb': {
      background: '#005A9E',

      [HighContrastSelector]: {
        forcedColorAdjust: 'none',
      },
    },

    ':active .ms-Slider-track': {
      background: '#005A9E',

      [HighContrastSelector]: {
        forcedColorAdjust: 'none',
      },
    },
  }),
});

/**
 * Styles for the rail slot
 */
const useRailStyles = makeStyles({
  rail: theme => ({
    height: '4px',
    width: '100%',
    background: '#C8C8C8',
    borderRadius: '4px',
    [HighContrastSelector]: {
      border: '1px solid WindowText',
    },
  }),

  vertical: theme => ({
    height: '120px',
    width: '4px',
  }),
});

/**
 * Styles for the track slot
 */
const useTrackStyles = makeStyles({
  track: theme => ({
    position: 'absolute',
    height: '4px',
    paddingRight: '2px', // Patches a minor gap between the rounded corners of the track and thumb
    background: '#606060',
    borderRadius: '4px 0px 0px 4px',

    [HighContrastSelector]: {
      background: 'Highlight',
    },
  }),

  vertical: theme => ({
    width: '4px',
    borderRadius: '0px 0px 4px 4px',
    paddingRight: '0px',
    paddingTop: '3px', // Patches a minor gap between the rounded corners of the track and thumb
  }),

  disabled: theme => ({
    // Still being worked on.
    background: 'red',

    [HighContrastSelector]: {
      background: 'Highlight',
    },
  }),
});

/**
 * Styles for the thumb slot
 */
const useThumbStyles = makeStyles({
  thumb: theme => ({
    position: 'absolute',
    flex: '1',
    width: '20px',
    height: '20px',
    background: '#606060',
    borderRadius: '50%',
    overflow: 'visible',
    // Doesn't work on Safari

    [HighContrastSelector]: {
      background: 'Highlight',
    },
  }),

  disabled: theme => ({
    // Theme still being worked on.
    background: 'red',
    [HighContrastSelector]: {
      background: 'Highlight',
    },
  }),

  focusIndicator: createFocusIndicatorStyleRule({
    outline: 'none',
    boxShadow: '0 0 0 1.5pt black',
    border: '1px solid #FFFFFF',

    [HighContrastSelector]: {
      background: 'GrayText',
      border: '2px solid WindowText',
    },
  }),
});

/**
 * Apply styling to the Slider slots based on the state
 */
export const useSliderStyles = (state: SliderState): SliderState => {
  const rootStyles = useRootStyles();
  const railStyles = useRailStyles();
  const trackStyles = useTrackStyles();
  const thumbStyles = useThumbStyles();

  state.className = mergeClasses(
    rootStyles.root,
    rootStyles.hover,
    rootStyles.focusWithin,
    rootStyles.activation,
    state.className,
  );

  state.rail.className = mergeClasses(railStyles.rail, state.rail.className);

  state.track.className = mergeClasses(
    trackStyles.track,

    state.track.className,
  );

  state.thumb.className = mergeClasses(
    thumbStyles.thumb,
    thumbStyles.focusIndicator,

    state.thumb.className,
  );

  return state;
};
