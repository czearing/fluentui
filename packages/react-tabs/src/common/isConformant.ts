import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath' | 'packageVersion'>) {
  const defaultOptions = {
    disabledTests: ['has-docblock', 'kebab-aria-attributes'],
    componentPath: module!.parent!.filename.replace('.test', ''),
    packageVersion: '@fluentui/react-tabs',
  };

  baseIsConformant(defaultOptions, testInfo);
}
