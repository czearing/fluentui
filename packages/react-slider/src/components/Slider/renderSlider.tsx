import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { SliderState } from './Slider.types';
import { sliderShorthandProps } from './useSlider';

/**
 * Render the final JSX of Slider
 */
export const renderSlider = (state: SliderState) => {
  const { slots, slotProps } = getSlots(state, sliderShorthandProps);

  return (
    <slots.root {...slotProps.root}>
      <slots.rail {...slotProps.rail} />
      <slots.track {...slotProps.track} />
      <slots.thumb {...slotProps.thumb} />
    </slots.root>
  );
};
