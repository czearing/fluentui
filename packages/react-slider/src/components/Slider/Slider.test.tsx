import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
import { safeCreate, safeMount } from '@fluentui/test-utilities';
import { resetIds } from '@fluentui/utilities';
import { Slider } from './Slider';
import { isConformant } from '../../common/isConformant';
/* eslint-disable @typescript-eslint/no-explicit-any */

describe('Slider', () => {
  beforeEach(() => {
    resetIds();
  });

  isConformant({
    Component: Slider,
    displayName: 'Slider',
    disabledTests: ['kebab-aria-attributes'],
  });

  it('renders Slider correctly', () => {
    safeCreate(<Slider defaultValue={5} />, component => {
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it('correctly handles (id) prop', () => {
    safeMount(<Slider id="test_id" />, component => {
      expect(component.find('.ms-Slider-root').getDOMNode().id).toEqual('test_id');
    });
  });

  it('correctly handles a (defaultValue) of zero', () => {
    let sliderRef: any;

    const SliderTestComponent = React.forwardRef((props, ref) => {
      sliderRef = React.useRef(null);

      return <Slider defaultValue={0} min={-100} max={100} ref={sliderRef} />;
    });

    safeMount(<SliderTestComponent />, component => {
      expect(sliderRef.current!.value).toEqual(0);
    });
  });

  it('correctly handles a (value) of zero', () => {
    let sliderRef: any;

    const SliderTestComponent = React.forwardRef((props, ref) => {
      sliderRef = React.useRef(null);

      return <Slider value={0} min={-100} max={100} ref={sliderRef} />;
    });

    safeMount(<SliderTestComponent />, component => {
      expect(sliderRef.current!.value).toEqual(0);
    });
  });

  it('correctly handles a (defaultValue) that is out of bounds', () => {
    let sliderRef: any;

    const SliderTestComponent = React.forwardRef((props, ref) => {
      sliderRef = React.useRef(null);

      return <Slider defaultValue={-10} min={0} max={100} ref={sliderRef} />;
    });

    safeMount(<SliderTestComponent />, component => {
      expect(sliderRef.current!.value).toEqual(0);
    });
  });

  it('correctly handles a (value) that is out of bounds', () => {
    let sliderRef: any;

    const SliderTestComponent = React.forwardRef((props, ref) => {
      sliderRef = React.useRef(null);

      return <Slider value={-10} min={0} max={100} ref={sliderRef} />;
    });

    safeMount(<SliderTestComponent />, component => {
      expect(sliderRef.current!.value).toEqual(0);
    });
  });
});
