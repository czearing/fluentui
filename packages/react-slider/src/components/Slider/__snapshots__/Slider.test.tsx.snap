// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders Slider correctly 1`] = `
<div
  className="ms-Slider-root"
  id="Slider22"
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="ms-Slider-rail"
  />
  <div
    className="ms-Slider-track"
    style={
      Object {
        "width": "50%",
      }
    }
  />
  <div
    aria-valuemax={10}
    aria-valuemin={0}
    aria-valuenow={5}
    aria-valuetext="5"
    className="ms-Slider-thumb"
    role="slider"
    style={
      Object {
        "left": "50%",
      }
    }
    tabIndex={0}
  />
</div>
`;
